Jupiter Notebook for Groupwork in AA 20224

Import Libaries

In [15]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from IPython.display import display
from datetime import timedelta

Load and Explore Data:

In [4]:
#data_path = ".\Data\"

charging_data = pd.read_csv(r'Data\\charging_sessions.csv')
weather_data = pd.read_csv(r'Data\weather_burbank_airport.csv')

try:
    print("charging info")
    display(charging_data.info())
except Exception as e:
    print("Error in charging_data.info():", e)

try:
    print("weather info")
    display(weather_data.info())
except Exception as e:
    print("Error in weather_data.info():", e)

try:
    print("charging head")
    display(charging_data.head())
except Exception as e:
    print("Error in charging_data.head():", e)

try:
    print("weather head")
    display(weather_data.head())
except Exception as e:
    print("Error in weather_data.head():", e)

try:
    print("charging description")
    display(charging_data.describe())
except Exception as e:
    print("Error in charging_data.describe():", e)

try:
    print("weather description")
    display(weather_data.describe())
except Exception as e:
    print("Error in weather_data.describe():", e)

charging info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


None

weather info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29244 entries, 0 to 29243
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   city                     29244 non-null  object 
 1   timestamp                29244 non-null  object 
 2   temperature              29219 non-null  float64
 3   cloud_cover              29224 non-null  float64
 4   cloud_cover_description  29224 non-null  object 
 5   pressure                 29236 non-null  float64
 6   windspeed                29158 non-null  float64
 7   precipitation            29244 non-null  float64
 8   felt_temperature         29218 non-null  float64
dtypes: float64(6), object(3)
memory usage: 2.0+ MB


None

charging head


Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

weather head


city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

charging description


Unnamed: 0  kWhDelivered        siteID        userID
count  66450.000000  66450.000000  66450.000000  49187.000000
mean    4726.102062     11.787916      1.472656   1800.494582
std     3568.025394     10.314789      0.499256   2617.548614
min        0.000000      0.501000      1.000000      1.000000
25%     1845.000000      5.097000      1.000000    431.000000
50%     3960.000000      9.140000      1.000000    818.000000
75%     7027.000000     14.183000      2.000000   1805.000000
max    15291.000000    108.797242      2.000000  19923.000000

weather description


temperature   cloud_cover      pressure     windspeed  precipitation  \
count  29219.000000  29224.000000  29236.000000  29158.000000   29244.000000   
mean      17.885622     30.107993    986.879400      8.566568       0.059360   
std        6.454604      5.400388      3.607065      6.699889       0.471682   
min        2.000000      4.000000    971.000000      0.000000       0.000000   
25%       13.000000     26.000000    984.500000      6.000000       0.000000   
50%       17.000000     33.000000    986.480000      7.000000       0.000000   
75%       22.000000     34.000000    989.110000     13.000000       0.000000   
max       46.000000     47.000000    999.650000     57.000000      18.540000   

       felt_temperature  
count      29218.000000  
mean          17.731433  
std            6.321100  
min            0.000000  
25%           13.000000  
50%           17.000000  
75%           22.000000  
max           42.000000

Check and handle missing Data

In [ ]:
##Missing Data
#Identify missing data
charging_missing_data = charging_data.isnull().sum()
display(charging_missing_data)

weather_missing_data = weather_data.isnull().sum()
display(weather_missing_data[weather_missing_data > 0])
print ("Number of unique lines with missing values: " + str(weather_data.isnull().any(axis=1).sum()))


Unnamed: 0              0
id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4088
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              17263
userInputs          17263
dtype: int64

temperature                25
cloud_cover                20
cloud_cover_description    20
pressure                    8
windspeed                  86
felt_temperature           26
dtype: int64

Number unique lines with missing values: 139


Lets look at the missing values:
For Charging data:
- doneChargingTime: We have 4088 missing values, that is 4088/66450=6,15%. Because we do not want to loose 6% of the data, we will replace the missing values with connectionTime + median charging time.
- UserID: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (0.0).
- UserInput: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (Null).

For Weather data:
- There are 139 lines with missing values throug the data set. Because this is only 0,47%, we decided to drop these records. 

In [70]:
#doneChargingTime 
#replace missing values with connectionTime + median
#store values from doneChargingTime into new column doneChargingTime_clean (can be transferred to doneChargingTime) in order to load them into a new csv

charging_data['connectionTime'] = pd.to_datetime(charging_data['connectionTime'])
charging_data['disconnectTime'] = pd.to_datetime(charging_data['disconnectTime'])
charging_data['doneChargingTime'] = pd.to_datetime(charging_data['doneChargingTime'])

#round median to whole second
median_charging_time = ((charging_data['doneChargingTime'] - charging_data['connectionTime']).dt.total_seconds()).median().round() 

doneChargingTime_median = charging_data['connectionTime'] + timedelta(seconds=median_charging_time)
charging_data['doneChargingTime_clean'] = charging_data['doneChargingTime'].fillna(doneChargingTime_median)


#check doneChargingTime_clean (can be deleted later)-->

#print(median_charging_time)

#print(doneChargingTime_median.iloc[1])
#print(charging_data['connectionTime'].iloc[1])

#comparison = charging_data[charging_data['doneChargingTime'].isnull()][['doneChargingTime', 'doneChargingTime_clean','disconnectTime']]
#print(comparison) <--



In [71]:
#userID
#replace missing values with 0.0 (float)
#store values from userID into new column userID_clean (can be transferred to userID)
charging_data['userID_clean'] = charging_data['userID'].fillna(0.0)

#check userID_clean for consistency (can be deleted later) -->
#userID_wrong_typ = charging_data['userID_clean'].apply(lambda x: not isinstance(x, float)).sum()
#userID_wrong_typ

#comparison = charging_data[charging_data['userID'].isnull()][['userID_clean', 'userID' ]]
#comparison['userID_clean_type'] = charging_data['userID_clean'].dtype
#print(comparison) 




Data Types 

In [75]:

#check for type of userID
userID_type = charging_data['userID'].dtype
print("data type of userID: ",userID_type)
#float64

#check for number of wrong  type within column userID
userID_wrong_type = charging_data['userID'].apply(lambda x: not isinstance(x, float)).sum()
print("number of incorrect data types: ", userID_wrong_type)
#-->no wrong types



data type of userID:  float64
number of incorrect data types:  0


Errorous Data,
Outliers,
Inconsistancies

In [ ]:
##Errorous Data
#Outliers
#Inconsistancies

#data['session_date'] = pd.to_datetime(data['session_date'])  # Convert to datetime

##Final Review

##data visualisation
#sns.histplot(data=charging_data, x='connectionTime', bins=30, kde=True)
#plt.title("Distribution of Charging Session Durations")
#plt.xlabel("Duration (minutes)")
#plt.ylabel("Frequency")
#plt.show()